zone-service/src/test/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneServiceTests.java

```diff
@@ -187,6 +181,53 @@ private OrchestratorZoneEntity buildOrchestratorZone() {
+                                                                          SUB_DOMAIN_NAME);
+        IdentityZone identityZone = createIdentityZone(null);
+        IdentityProvider identityProvider = createDefaultIdp(identityZone);
+        when(zoneProvisioning.create(any())).thenReturn(identityZone);
+        when(idpProvisioning.create(any(),any())).thenReturn(identityProvider);
+        when(clientDetailsService.create(any(),any())).thenThrow(new OrchestratorZoneServiceException("Client Already exists"));
+        Assertions.assertThrows(OrchestratorZoneServiceException.class, () -> zoneService.createZone(zoneRequest),
+                                "Client Already exists exception not thrown");
+        verify(idpProvisioning, times(1)).create(any(),any());
+        verify(clientDetailsService, times(1)).create(any(),any());
+        verify(clientDetailsValidator, times(1)).validate(any(),any());
+    }
+
+    @Test
+    public void testCreateZone_createDefaultIdp_ExceptionCheck() throws OrchestratorZoneServiceException {
+        Security.addProvider(new BouncyCastleProvider());
+        OrchestratorZoneRequest zoneRequest =  getOrchestratorZoneRequest(ZONE_NAME, ADMIN_CLIENT_SECRET,
+                                                                          SUB_DOMAIN_NAME);
+        IdentityZone identityZone = createIdentityZone(null);
+        IdentityProvider identityProvider = createDefaultIdp(identityZone);
+        when(zoneProvisioning.create(any())).thenReturn(identityZone);
+        when(idpProvisioning.create(any(),any())).thenThrow(new IdpAlreadyExistsException("IDP Already exists"));
+        when(clientDetailsService.create(any(),any())).thenReturn(any());
+
+        Assertions.assertThrows(OrchestratorZoneServiceException.class, () -> zoneService.createZone(zoneRequest),
+                                "IDP Already exists exception not thrown");
+    }
+
+    @Test
+    public void testCreateZone_createIdentityZone_ExceptionCheck() throws OrchestratorZoneServiceException {
+        Security.addProvider(new BouncyCastleProvider());
+        OrchestratorZoneRequest zoneRequest =  getOrchestratorZoneRequest(ZONE_NAME, ADMIN_CLIENT_SECRET,
+                                                                          SUB_DOMAIN_NAME);
+        IdentityZone identityZone = createIdentityZone(null);
+        IdentityProvider identityProvider = createDefaultIdp(identityZone);
+        when(zoneProvisioning.create(any())).thenThrow(new ZoneAlreadyExistsException("Identity Zone Already exists"));
+        when(idpProvisioning.create(any(),any())).thenReturn(identityProvider);
+        when(clientDetailsService.create(any(),any())).thenReturn(any());
+
+        Assertions.assertThrows(ZoneAlreadyExistsException.class, () -> zoneService.createZone(zoneRequest),
+                                "Identity Zone Already exists exception not thrown");
+    }
+
```

Comment:
OrchestratorZoneService::generateIdentityZone() can throw two checked exceptions:

- **IOException** that might be caused by createSigningKey() and createSamlConfig()
- **NoSuchAlgorithmException** that might be caused by createSamlConfig().  

Are you wrapping those in OrchestratorZoneServiceException? Can you test?

Reply:
As I mentioned in the subtasks, we cannot test generateIdentityZone method because all the code completely private to the method, unless we remove RSA algorithm or some JVM changes we cannot test this.

The above one testing exception thrown inside createIdentityZone method.

Reply:
As discussed, there is no transactions happening at the time generateIdentityZone() is called.  No need to validate that exceptions caused by generateIdentityZone() are runtime.